llvm/lib/SYCLLowerIR/LowerWGLocalMemory.cpp

```diff
@@ -98,8 +98,8 @@ static bool allocaWGLocalMemory(Module &M) {
 
   assert(ALMFunc->isDeclaration() && "should have declaration only");
 
-  for (User *U : ALMFunc->users()) {
-    auto *CI = cast<CallInst>(U);
+  for (auto U = ALMFunc->user_begin(), UE = ALMFunc->user_end(); U != UE;) {
```

Comment:
I think a conventional way to do this is to employ a worklist of instructions and then iterate over it without worrying about invalidating iterators.

Reply:
Another possible approach is to collect instructions which you need to remove to some container and remove them later in a separate loop:

https://github.com/intel/llvm/blob/691f8428ad87c27136e6aa2a01729b86d485c03e/llvm/tools/sycl-post-link/SpecConstants.cpp#L622-L626